snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclEvaluationRequest.java

```diff
@@ -646,55 +647,40 @@ protected Class<?> getDocumentType() {
 	}
 
 	@Override
-	protected void addParentIds(IComponent concept, Set<String> collection) {
-		if (concept instanceof SnomedConcept) {
-			SnomedConcept snomedConcept = (SnomedConcept) concept;
-			if (Trees.INFERRED_FORM.equals(expressionForm)) {
-				if (snomedConcept.getParentIds() != null) {
-					for (long parent : snomedConcept.getParentIds()) {
-						if (IComponent.ROOT_IDL != parent) {
-							collection.add(Long.toString(parent));
-						}
-					}
-				}
-			} else {
-				if (snomedConcept.getStatedParentIds() != null) {
-					for (long statedParent : snomedConcept.getStatedParentIds()) {
-						if (IComponent.ROOT_IDL != statedParent) {
-							collection.add(Long.toString(statedParent));
-						}
-					}
-				}
-			}
+	protected Collection<String> getParentIds(IComponent concept) {
+		if (!(concept instanceof SnomedConcept snomedConcept)) {
+			return super.getParentIds(concept);
+		}
+		
+		if (Trees.INFERRED_FORM.equals(expressionForm)) {
+			return toStringSet(snomedConcept.getParentIds());
 		} else {
-			super.addParentIds(concept, collection);
+			return toStringSet(snomedConcept.getStatedParentIds());
 		}
 	}
-	
+
 	@Override
-	protected void addAncestorIds(IComponent concept, Set<String> collection) {
-		if (concept instanceof SnomedConcept) {
-			SnomedConcept snomedConcept = (SnomedConcept) concept;
-			if (Trees.INFERRED_FORM.equals(expressionForm)) {
-				if (snomedConcept.getAncestorIds() != null) {
-					for (long ancestor : snomedConcept.getAncestorIds()) {
-						if (IComponent.ROOT_IDL != ancestor) {
-							collection.add(Long.toString(ancestor));
-						}
-					}
-				}
-			} else {
-				if (snomedConcept.getStatedAncestorIds() != null) {
-					for (long statedAncestor : snomedConcept.getStatedAncestorIds()) {
-						if (IComponent.ROOT_IDL != statedAncestor) {
-							collection.add(Long.toString(statedAncestor));
-						}
-					}
-				}
-			}
+	protected Collection<String> getAncestorIds(IComponent concept) {
+		if (!(concept instanceof SnomedConcept snomedConcept)) {
+			return super.getAncestorIds(concept);
+		}
+
+		if (Trees.INFERRED_FORM.equals(expressionForm)) {
+			return toStringSet(snomedConcept.getAncestorIds());
 		} else {
-			super.addAncestorIds(concept, collection);
+			return toStringSet(snomedConcept.getStatedAncestorIds());
+		}
+	}
+	
+	private Set<String> toStringSet(final long[] idArray) {
+		if (idArray == null || idArray.length < 1) {
+			return newHashSet();
```

Comment:
I'm just wondering if this should be mutable. Or would it be possible to use an immutable empty set here?